041-opa-integration/proposal.md

```diff
@@ -0,0 +1,147 @@
+# Summary
+
+This proposal outlines the beginnings of support for policy enforcement. OPA 
+(Open Policy Agent) will be used as the first policy manager to integrate with.
+
+
+# Motivation
+
+Generally speaking, anything that exposes an HTTP API (whether an individual 
+micro-service or an application as a whole) needs to control who can run those 
+APIs and when. 
+
+Concourse, as a CI platform, also needs to control who can do something and when,
+which can be done by integrating with a policy manager. Before any action, send
+a policy check request, with data that describes the action, to the policy manager,
+and only continue if the policy manager replies a "pass".
+
+Possible policies to apply could be:
+
+* Steps, e.g. code security scan, by certain resource types must be done.
+* Some team's tasks must be run with certain tag.
+* Some docker image registries are not allowed.
+* Some docker images are not not allowed.
+* and so on ...
+
+
+# Proposal
+
+## Support multiple policy managers
+
+Like how multiple credential managers are supported, Concourse should allow other
+policy managers than OPA.
+
+## Policy check points
+
+* All API calls, for example `set-pipeline`, will also go through policy checks.
+* Because Concourse supports task file, and task file is only fetched at runtime.
+The action `set-pipeline` has no access to task files. To support policy check 
+against tasks, a policy check should be invoked before a task run starts.
+
+## OPA configuration
+
+* CONCOURSE_OPA_URL - URL of OPA service including path
```

Comment:
Do you see this as something that should only be configured cluster-wide? Or would it make sense to you to have this configured on a per-team basis?

Reply:
For our use cases, we don't need per-team OPA. If a team/pipeline wants to use their own OPA, an OPA resource type can be used.